Fees and Rewards (#25)
subspace/protocol-specs: docs/fees_and_rewards/Dynamic_Issuance.md

---
title: Dynamic Issuance
sidebar_position: 3
description: Dynamic Issuance Formulas
keywords:
    - fees
    - rewards
    - WIP
last_update:
  date: 04/16/2024
  author: Saeid Yazdinejad
---
import Collapsible from '@site/src/components/Collapsible/Collapsible';

# Dynamic Issuance
As opposed to having a fixed issuance rate, the Subspace Network implements a dynamic issuance rate based on network usage, where, roughly, the inflation rate is inversely proportional to the utilization rate of the blockspace. 

TLDR: The farmer who proposed a block gets some fresh SSC + fees, and voters get some fresh SSC regardless of what the proposer got.

> [Dynamic Issuance Specification](https://www.notion.so/Dynamic-Issuance-Specification-3bdf63955cb943489347889775d71c24?pvs=21)

Reviewer: Let's avoid linking to Notion. This current md page will supersede the notion page.

# Space Race

> The rewards issuance is not enabled by default at the chain genesis. After the genesis block, several things need to happen before authoring blocks by anyone, and  [rewards can be enabled](https://subnomicon.subspace.network/docs/consensus/genesis). Then:

Reviewer: ```suggestion
The rewards issuance is not enabled by default at the chain genesis. After the genesis block, several things need to happen before authoring blocks by anyone, and rewards can be enabled (See [genesis](https://subnomicon.subspace.network/docs/consensus/genesis) algorithm). Then:
```

> 1. Define a target space pledged to the network at which the rewards will be enabled (i.e., 8PiB for Gemini-3h). In the runtime, it is defined in terms of the target value of the solution range [`solution_range_for_rewards`](https://github.com/subspace/protocol-specs/issues/13).

Reviewer: Please add a TODO to update this link.

2. Identify an assumed current space pledged `initial_solution_range`.
3. Enable solution range adjustment starting at `initial_solution_range` at the next era 
4. Set in the runtime the `solution_range_for_rewards` below which to automatically enable rewards.
5. Enable block authoring by anyone.

> 6. Farmers will start producing blocks and the solution range will be updated every era as defined in [Solution Range Updates](https://www.notion.so/Solution-Range-Updates-a7205a46a20c42e5838d0bc332c51bed?pvs=21) 

Reviewer: All notion links should be replaced with relevant links to the specs or issues filed when the page/paragraph doesn't exist. This particular should link to `docs/consensus/proof_of_archival_storage.md#solution-range-updates`

7. Eventually, as farmers pledge more space, the network’s solution range will reach a value ≤ `solution_range_for_rewards` at a certain block $B$.
8. The `RewardsEnabled` runtime item is set to `true`. The rewards will start to be paid to voters and proposer of the very next block.

# Issuance Function

The issuance function currently consists of two components: a decaying component and a utilization-based component. The decaying component issues a constant reward for some initial interval and gradually reduces it over time. The utilization-based component issues a reward based on the average utilization of blockspace in recent blocks.
Such construction allows for more decaying components to be included for additional incentivization.

> **Full  Block**

Reviewer: Please make all of these section titles ### instead of bold face. This way it will be possible to directly link to specific sections from other places.

4. Else if `block_height <= num_blocks`, store new value 
`(AvgBlockspaceUsage + used_blockspace)/2`
5. After `block_height > num_blocks`, use `multiplier = 2/(num_blocks+1)`
Store new value 
`multiplier * used_blockspace + (1-multiplier) * AvgBlockspaceUsage`

**Issuance Components**

We define a base block issuance as the number of tokens issued per block in a static case: a fixed reward to block proposer and `EXPECTED_VOTES_PER_BLOCK` voters.

`RemainingIssuance`: tokens left to issue to farmers at any given time

We define a reward split rule where the block proposer and voters are entitled to a share of the base reward:

- `proposer_share = 1`
- `voters_share`, currently `=EXPECTED_VOTES_PER_BLOCK` (for gemini-3h)
- `total_shares = proposer_share + voters_share`

The total actual reward issued includes a sum of several independent issuance components, each characterized by different values of common parameters based on the value of `BASE_REWARD` computed off-chain beforehand as described in [Off-chain Issuance curve parameters setting](https://www.notion.so/Off-chain-Issuance-curve-parameters-setting-1e532d57bcff4475933c395919505c85?pvs=21).  

> $B$ is the block starting from which the rewards component is activated. It can be set manually by an extrinsic initializing the component or automatically if rewards are subject to [Space Race](https://www.notion.so/Space-Race-9ea6d89b884c422a98fcbe3562a63e36?pvs=21) and stored in `RewardsEnabled` runtime item.

Reviewer: This should link to Space race section above.

Defining a component requires setting the following parameters:

`struct RewardPoint`:

A reward point is a tuple of `(block, subsidy)`. Each point is a start block of a rewards decay phase and the subsidy amount for that block. For blocks between each pair of points, we use linear approximation to compute their respective subsidies.

Block numbers have to be strictly increasing, and reward values have to be strictly decaying across all phases.

The length of this list is the number of decay phases, usually 4-6. 

All block numbers will be offset by $B$ at the time the component is initialized or as soon as $B$ is known if rewards are subject to [Space Race](https://www.notion.so/Space-Race-9ea6d89b884c422a98fcbe3562a63e36?pvs=21).

We currently define two such components: block proposer reference subsidy and voter reward. If desired, we may introduce other components (i.e., one to incentivize operators for an initial period of low domain activity).

(Numbers for Gemini-3h):

`ProposerSubsidyPoints`: vector of `RewardPoint`s

`[(0, 100000000000000000), (201600, 99989921015995728), (79041600, 92408728791312960), (779041600, 45885578019877912), (2443104160, 8687806947398648)]`

> The first value of `subsidy` here is equal to `proposer_share/total_shares*BASE_REWARD`, and the rest follow exponential decay, computed off-chain as described in [Off-chain Issuance curve parameters setting](https://www.notion.so/Off-chain-Issuance-curve-parameters-setting-1e532d57bcff4475933c395919505c85?pvs=21). The points describe the following phases:

Reviewer: This should link to the section below. Same in line 116 and 80

</Collapsible>

`block_reward(block_height) → Balance`

The reward to be issued to the proposer for this block is defined by a reward function and computed for each block using latest `AvgBlockspaceUsage`, `transaction_byte_fee` and `block_height`.

```rust
reference_subsidy-blockspace_utilization*min(reference_subsidy,max_block_fee)
```
where

```rust
reference_subsidy = reference_subsidy_for_block(ProposerSubsidyParams, block_height)

max_block_fee = MAX_NORMAL_BLOCK_LENGTH * transaction_byte_fee

blockspace_utilization=AvgBlockspaceUsage/MAX_NORMAL_BLOCK_LENGTH
```

<Collapsible title="Original Formula Simplification Note">
The above definition is an approximation of the following hyperbolic formula for the chosen parameters below:
<center>
$a+b\tanh(-c(\text{blockspace\_utilization}-d))$
</center>

where 

- $a =S_r - b*\tanh(c*d)$  the offset parameter (sets the amount of reward issued at 0 utilization to $S_r$).

> - $~~S_r~~$ is `reference_subsidy = reference_subsidy_for_block(ProposerSubsidyParams, block_height)`, a maximum amount of SSC issued at 0 utilization.

Reviewer: ```suggestion
- $S_r$ is `reference_subsidy = reference_subsidy_for_block(ProposerSubsidyParams, block_height)`, a maximum amount of SSC issued at 0 utilization.
```

> - $b$  $= \frac{S_r-\max(S_r-\bar{F},0)}{\text{const}\tanh(c*d)}=\frac{\min(S_r,\bar{F})}{\text{const}\tanh(c*d)}$ is a linear sensitivity parameter.

Reviewer: ```suggestion
- $b = \frac{S_r-\max(S_r-\bar{F},0)}{\text{const}\tanh(c*d)}=\frac{\min(S_r,\bar{F})}{\text{const}\tanh(c*d)}$ is a linear sensitivity parameter.
```

    <Collapsible title="Note">
        c is a parameter that controls how slow will rewards go down as utilization ratio goes up. This is useful for not letting the rewards go too down when utilization ratios are small. The higher the values for $c$, the slower the rewards will decay as the utilization ratio goes up. Values near zero imply in a straight line from rewards being reference_subsidy(utilization=0) until zero (utilization=1). Values around ~1.0 imply in a somewhat curved line from rewards being reference_subsidy(g=0) until zero (g=1). For instance, when g=0.5, the rewards would be around ~20% larger than compared with the first case. As values goes way above 1.0, then the result is to have a relatively constant reward for low values of utilization ratio, with a exponential decay afterwards. Generally speaking, Values between 0 and 1 imply that it is rational for farmers to try to maximize utilization ratio. For values above 1.0 that's not obvious, as the peak profit (rewards + fees) may be maximized on inflection point (generally between 60-80% of the utilization ratio)
    </Collapsible> 

> - $\text{blockspace\_utilization}$ = `AvgBlockspaceUsage/MAX_NORMAL_BLOCK_LENGTH`

Reviewer: This doesn't seem to render correctly on the frontend

Author: ![Screenshot 2024-04-16 at 5 30 44 PM](https://github.com/subspace/protocol-specs/assets/89309559/0424130a-8271-4193-bcfc-e4f553e05bd3)

I check it out again. What is not correct here? 🤔

Reviewer: Hmm looks like it's Github only issue then 
<img width="618" alt="Screenshot 2024-04-17 at 10 22 38" src="https://github.com/subspace/protocol-specs/assets/35382599/646c5952-d33a-4b3b-b4aa-f688f55c7d25">

Author: I changed it to this format:
blockspace_utilization = `AvgBlockspaceUsage/MAX_NORMAL_BLOCK_LENGTH`

- (const) $d=1$, utilization rate at which reward issued is 0

Putting everything together:

<center>
$
a+b\tanh(-c(\text{blockspace\_utilization}-d))=\\S_r - b*\tanh(c*d) + b\tanh(-c(\text{blockspace\_utilization}-d))=\\S_r-b(\tanh(c*d)-\tanh(-c(\text{blockspace\_utilization}-d)))\approx \\ |c=0.99,d=1|\approx \\S_r-b(\tanh(c*d)-\tanh(c*d)(1-\text{blockspace\_utilization}))=\\S_r-b\tanh(c*d)\text{blockspace\_utilization}=\\|b=\frac{\min(S_r,\bar F)}{\tanh(c*d)}|=\\S_r-\text{blockspace\_utilization}*\min(S_r,\bar F)
$
</center>

</Collapsible>

**Voting Reward**

By vote rewards we mean SSC newly issued by the protocol to incentivize voting to each voter.

> Vote rewards have  *a probabilistic issuance rate* based on expected number of votes in the block. **Reward **numbers **defined in `VoterSubsidyParams` are issued **per vote** with all voters getting equal reward.

Reviewer: ```suggestion
Vote rewards have  *a probabilistic issuance rate* based on expected number of votes in the block. Reward numbers defined in `VoterSubsidyParams` are issued **per vote** with all voters getting equal reward.
```

`num_blocks`: number of the blocks over which utilization is measured, should be ≥0 (Currently, 100)

1. Compute the amount of blockspace utilized in this block `used_blockspace` ****as the sum of encoded lengths of normal transactions in block body.
2. For genesis block `AvgBlockspaceUsage = used_blockspace`=0
3. If `num_blocks==0` , return the `used_blockspace` of the current block.
4. Else if `block_height <= num_blocks`, store new value 
`(AvgBlockspaceUsage + used_blockspace)/2`
5. After `block_height > num_blocks`, use `multiplier = 2/(num_blocks+1)`
Store new value 
`multiplier * used_blockspace + (1-multiplier) * AvgBlockspaceUsage`

### Issuance Components

We define a base block issuance as the number of tokens issued per block in a static case: a fixed reward to block proposer and `EXPECTED_VOTES_PER_BLOCK` voters.

`RemainingIssuance`: tokens left to issue to farmers at any given time

We define a reward split rule where the block proposer and voters are entitled to a share of the base reward:

- `proposer_share = 1`
- `voters_share`, currently `=EXPECTED_VOTES_PER_BLOCK` (for gemini-3h)
- `total_shares = proposer_share + voters_share`

> The total actual reward issued includes a sum of several independent issuance components, each characterized by different values of common parameters based on the value of `BASE_REWARD` computed off-chain beforehand as described in [Off-chain Issuance curve parameters setting](docs/fees_and_rewards/Dynamic_Issuance#off-chain-issuance-curve-parameters-setting).  

Reviewer: ```suggestion
The total actual reward issued includes a sum of several independent issuance components, each characterized by different values of common parameters based on the value of `BASE_REWARD` computed off-chain beforehand as described in [Off-chain Issuance curve parameters setting](#off-chain-issuance-curve-parameters-setting).  
```

> $B$ is the block starting from which the rewards component is activated. It can be set manually by an extrinsic initializing the component or automatically if rewards are subject to [Space Race](docs/fees_and_rewards/Dynamic_Issuance#space-race) and stored in `RewardsEnabled` runtime item.

Reviewer: I realized this was not accurate.
```suggestion
$B$ is the block starting from which the rewards component is activated. It can be set manually by an extrinsic initializing the component or automatically if rewards are subject to [Space Race](#space-race).
```

Defining a component requires setting the following parameters:

`struct RewardPoint`:

A reward point is a tuple of `(block, subsidy)`. Each point is a start block of a rewards decay phase and the subsidy amount for that block. For blocks between each pair of points, we use linear approximation to compute their respective subsidies.

Block numbers have to be strictly increasing, and reward values have to be strictly decaying across all phases.

The length of this list is the number of decay phases, usually 4-6. 

> All block numbers will be offset by $B$ at the time the component is initialized or as soon as $B$ is known if rewards are subject to [Space Race](docs/fees_and_rewards/Dynamic_Issuance#space-race).

Reviewer: ```suggestion
If a component is initialized before the rewards are activated via [Space Race](#space-race) (i.e. at genesis), all block heights in the reward points will be offset by the block number $B$ at which rewards get activated as soon as Space Race completes. Else, if a new component is defined over pre-existing rewards, the block heights in the reward points list are offset by the block number at which the component initialization network upgrade goes through.
```

We currently define two such components: block proposer reference subsidy and voter reward. If desired, we may introduce other components (i.e., one to incentivize operators for an initial period of low domain activity).

(Numbers for Gemini-3h):

`ProposerSubsidyPoints`: vector of `RewardPoint`s

`[(0, 100000000000000000), (201600, 99989921015995728), (79041600, 92408728791312960), (779041600, 45885578019877912), (2443104160, 8687806947398648)]`

> The first value of `subsidy` here is equal to `proposer_share/total_shares*BASE_REWARD`, and the rest follow exponential decay, computed off-chain as described in [Off-chain Issuance curve parameters setting](docs/fees_and_rewards/Dynamic_Issuance#off-chain-issuance-curve-parameters-setting). The points describe the following phases:

Reviewer: ```suggestion
The first value of `subsidy` here is equal to `proposer_share/total_shares*BASE_REWARD`, and the rest follow exponential decay, computed off-chain as described in [Off-chain Issuance curve parameters setting](#off-chain-issuance-curve-parameters-setting). The points describe the following phases:
```

- decay phase 0:  `0` ≤ block height < `201600`, subsidy `100000000000000000 -> 99989921015995728`
- decay phase 1: `201600` ≤ block height < `79041600`, subsidy `99989921015995728 -> 92408728791312960` decay phase 1
- decay phase 2: `79041600` ≤ block height < `779041600`, subsidy `92408728791312960 -> 45885578019877912` decay phase 2
- decay phase 3: `779041601` ≤ block height < `2443104160`, subsidy `92408728791312960 ->8687806947398648)]`
- decay phase 4: `2443104160` ≤ block height, constant subsidy of `8687806947398648` until remaining issuance runs out.

`VoterSubsidyPoints`: vector of `RewardPoint`s

`[(0, 100000000000000000), (201600, 99989921015995728), (79041600, 92408728791312960), (779041600, 45885578019877912), (2443104160, 8687806947398648)]`

> The first value `from_subsidy` is the initial subsidy equal to `voter_share/total_shares*BASE_REWARD`, and the rest follow exponential decay, computed off-chain as described in [Off-chain Issuance curve parameters setting](docs/fees_and_rewards/Dynamic_Issuance#off-chain-issuance-curve-parameters-setting) 

Reviewer: ```suggestion
The first value `from_subsidy` is the initial subsidy equal to `voter_share/total_shares*BASE_REWARD`, and the rest follow exponential decay, computed off-chain as described in [Off-chain Issuance curve parameters setting](#off-chain-issuance-curve-parameters-setting) 
```
